README.md
.. image:: https://matthieumeo.github.io/hvox/html/_images/hvox.png
  :width: 50 %
  :align: center
  :target: https://matthieumeo.github.io/hvox/html/index

[![License BSD-3](https://img.shields.io/pypi/l/hvox.svg?color=green)](https://github.com/matthieumeo/hvox/raw/main/LICENSE)
[![PyPI](https://img.shields.io/pypi/v/hvox.svg?color=green)](https://pypi.org/project/hvox)
[![Python Version](https://img.shields.io/pypi/pyversions/hvox.svg?color=green)](https://python.org)

Python reference implementation of the HVOX gridder for mesh-agnostic wide-field interferometry [1].

----------------------------------

## Installation

    # Create conda environment 
    my_env=<CONDA ENVIRONMENT NAME>
    conda create --name "${my_env}"
    
    # Install HVOX
    python -m pip install git+https://github.com/matthieumeo/hvox" 

    # Install additional dependencies to run examples and CLI  
    python -m pip install --index-url=https://artefact.skao.int/repository/pypi-all/simple rascil
    python -m pip install healpy

    # Developer Install HVOX
    python -m pip install "hvox[dev] @ git+https://github.com/matthieumeo/hvox" 

## Contributing

Contributions are very welcome. Tests can be run with [tox], please ensure
the coverage at least stays the same before you submit a pull request.

## License

Distributed under the terms of the [MIT] license,
`hvox` is free and open source software

## Issues

If you encounter any problems, please [file an issue] along with a detailed description.

[MIT]: http://opensource.org/licenses/MIT

[file an issue]: https://github.com/matthieumeo/hvox/issues

[tox]: https://tox.readthedocs.io/en/latest/
[pip]: https://pypi.org/project/pip/
[PyPI]: https://pypi.org/

## References

[1] [HVOXpaper]Kashani, S., Queralt, J. R., Jarret, A., & Simeoni, M. (2023). HVOX: Scalable Interferometric Synthesis and Analysis of Spherical Sky Maps. arXiv [Cs.CE]. Retrieved from http://arxiv.org/abs/2306.06007

ChangeLog
CHANGES
=======

* correct reference
* add citation file
* update installation setup and instructions
* Revert "update authors"
* update authors
* ignore tart-related files
* [bugfix] update nufft call after NUFFT args update
* update dependencies
* update changelog
* add healpix example
* ref. and add support for mesh=="dcos"
* remove hpx example (tmp)
* [tests][BugFix] change epsilon to 2x requested (1x per method)
* [tests][BugFix] remove 1/Nvis factor to match ducc0 result
* remove mesh argument (tmp)
* add phase center precision
* add symmetric baselines examples
* modify requirements
* add phase center in example
* add installation instructions
* add psf-center
* add mesh=dcos option
* [changelog] update
* [build] add pyproject.toml
* [cli] add ms2fits
* [setup] add cli
* [setup] update pycsou requirement
* [example\_hpix] update WIP
* [example\_hpix] add WIP
* [test][core] change epsilon for w-term=False
* [tests][core][BugFix] test dirty images too
* [core][doc] homogenize
* [changelog][auto-update]
* [doc] update index and references
* [doc] add methods description
* [authors/changelog][auto-update]
* [example][BugFix] add visibility weights / mask
* [tests][core] add test for main functions
* [core][BugFix] add visibility weights / mask
* [example\_hpix] add data
* [example\_dcos] use generic separation (os.sep)
* [example\_dcos] change print
* [example\_dcos][bugfix] change arg\_name
* [tests] add test structure
* [DOC][BugFix] add correct url for pip-installation
* Add content [setup, readme, doc, src, data, example]
* Initial commit


src/hvox_tests/test_core.py
import numpy as np

try:
    import astropy.units as u
    import ducc0.wgridder.experimental as wgridder
    import pytest
    from astropy.coordinates import SkyCoord
    from rascil.processing_components import create_test_image, show_image
    from ska_sdp_datamodels.configuration.config_create import (
        create_named_configuration,
    )
    from ska_sdp_datamodels.science_data_model.polarisation_model import (
        PolarisationFrame,
    )
    from ska_sdp_datamodels.visibility import create_visibility
    from ska_sdp_func_python.util.coordinate_support import skycoord_to_lmn
except:
    raise ValueError("Missing packages, install `test` requirements.")

import hvox


class TestHVOX_dcos:
    @pytest.fixture
    def gt(self):
        lowr3 = create_named_configuration("LOWBD2", rmax=3_000.0)
        vis = create_visibility(
            config=lowr3,
            times=np.linspace(-4, 4, 9) * np.pi / 12,
            frequency=np.r_[15e7],
            channel_bandwidth=np.r_[5e4],
            weight=1.0,
            phasecentre=SkyCoord(
                ra=+15.0 * u.deg, dec=-45.0 * u.deg, frame="icrs", equinox="J2000"
            ),
            polarisation_frame=PolarisationFrame("stokesI"),
            times_are_ha=False,
        )

        print("Loading sky GT")
        model = create_test_image(
            phasecentre=vis.phasecentre, frequency=np.r_[15e7], cellsize=5e-4
        )

        direction_cosines, jacobian = get_direction_cosines(model)
        wgt_dirty = 1 / jacobian.reshape(-1)

        return vis, model, direction_cosines, wgt_dirty

    @pytest.fixture(params=["single", "double"])
    def dtype(self, request):
        dtype_r = np.single if request.param == "single" else np.double
        dtype_c = np.csingle if request.param == "single" else np.cdouble
        return [dtype_r, dtype_c]

    @pytest.fixture
    def vis(self, gt, dtype):
        visibilities = gt[0].copy()
        visibilities.vis.data = visibilities.vis.data.astype(dtype[1])
        return visibilities

    @pytest.fixture
    def model(self, gt, dtype):
        sky_model = gt[1]
        sky_model.pixels.data = sky_model.pixels.data.astype(dtype[0])
        return sky_model

    @pytest.fixture
    def direction_cosines(self, gt):
        return gt[2]

    @pytest.fixture
    def wgt_dirty(self, gt, dtype):
        return gt[3].astype(dtype[0])

    @pytest.fixture(params=["w-True", "w-False"])
    def w_term(self, request):
        if request.param == "w-True":
            return True
        else:
            return False

    @pytest.fixture
    def epsilon(self, dtype, w_term):
        if w_term:
            if np.issubdtype(dtype[0], np.single):
                return 1e-5
            else:
                return 1e-9
        else:
            return 1e-3

    @pytest.fixture(params=["ch-True", "ch-False"])
    def chunked(self, request):
        if request.param == "ch-True":
            return True
        else:
            return False

    @pytest.fixture
    def hvox_results(
        self,
        vis,
        direction_cosines,
        model,
        wgt_dirty,
        w_term,
        epsilon,
        chunked,
    ):
        vis_hvox = hvox.dirty2vis(uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3),
                                  xyz=direction_cosines.reshape(-1, 3), dirty=model.pixels.data.reshape(-1),
                                  w_term=w_term, epsilon=epsilon, chunked=chunked)

        dirty_hvox = hvox.vis2dirty(uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3),
                                    xyz=direction_cosines.reshape(-1, 3), vis=vis_hvox.reshape(-1), w_term=w_term,
                                    epsilon=epsilon, chunked=chunked)
        return vis_hvox, dirty_hvox

    @pytest.fixture
    def ducc0_results(
        self,
        vis,
        model,
        epsilon,
        w_term,
    ):
        npixels = model.pixels.shape[-1]
        pixsize = np.abs(np.radians(model.image_acc.wcs.wcs.cdelt[0]))
        fuvw = vis.uvw.data.copy().reshape(-1, 3)
        fuvw[:, 0] *= -1.0
        fuvw[:, 2] *= -1.0
        vis_ducc0 = wgridder.dirty2vis(
            uvw=fuvw,
            freq=vis.frequency.data,
            dirty=model.pixels.data.squeeze().T,
            wgt=vis.visibility_acc.flagged_weight.reshape(-1, 1).astype(
                model.pixels.data.dtype
            ),
            pixsize_x=pixsize,
            pixsize_y=pixsize,
            epsilon=epsilon,
            do_wgridding=w_term,
            nthreads=4,
            verbosity=0,
        )

        dirty_ducc0 = wgridder.vis2dirty(
            uvw=fuvw,
            freq=vis.frequency.data,
            vis=vis_ducc0.reshape(-1, 1),
            npix_x=npixels,
            npix_y=npixels,
            pixsize_x=pixsize,
            pixsize_y=pixsize,
            epsilon=epsilon,
            do_wgridding=w_term,
            nthreads=4,
            verbosity=0,
        ).T
        return vis_ducc0, dirty_ducc0

    def test_dirty2vis(self, ducc0_results, hvox_results, epsilon):
        assert np.linalg.norm(
            ducc0_results[0].ravel() - hvox_results[0].ravel()
        ) / np.linalg.norm(ducc0_results[0]) < 2 * (epsilon * 10)

    def test_vis2dirty(self, ducc0_results, hvox_results, epsilon):
        assert np.linalg.norm(
            ducc0_results[1].ravel() - hvox_results[1].ravel()
        ) / np.linalg.norm(ducc0_results[1]) < 2 * (epsilon * 10)


def get_direction_cosines(image):
    _, _, _, npixel = image.pixels.data.shape
    _, _, ny, nx = image["pixels"].shape
    lmesh, mmesh = np.meshgrid(np.arange(ny), np.arange(nx))
    ra_grid, dec_grid = image.image_acc.wcs.sub([1, 2]).wcs_pix2world(lmesh, mmesh, 0)
    ra_grid = np.deg2rad(ra_grid)
    dec_grid = np.deg2rad(dec_grid)
    directions = SkyCoord(
        ra=ra_grid.ravel() * u.rad,
        dec=dec_grid.ravel() * u.rad,
        frame="icrs",
        equinox="J2000",
    )
    l, m, _ = skycoord_to_lmn(directions, image.image_acc.phasecentre)
    jacobian = np.sqrt(1 - l**2 - m**2)
    direction_cosines = np.stack([l, m, jacobian - 1.0], axis=-1).reshape(-1, 3)
    return direction_cosines, jacobian


example_hpix.py
import os

import matplotlib.pyplot as plt
import numpy as np

try:
    import astropy.units as u
    import astropy_healpix
    import healpy as hp
    from astropy.coordinates import Galactic, SkyCoord
    from astropy.io import fits
    from astropy_healpix import HEALPix
    from rascil.processing_components import create_test_image, show_image
    from ska_sdp_datamodels.configuration.config_create import (
        create_named_configuration,
    )
    from ska_sdp_datamodels.science_data_model.polarisation_model import (
        PolarisationFrame,
    )
    from ska_sdp_datamodels.visibility import create_visibility
    from ska_sdp_func_python.util.coordinate_support import skycoord_to_lmn
except:
    raise ValueError("Missing packages")

import hvox

if __name__ == "__main__":

    # Sparse or dense sky
    sparse_sky = False

    # Radius of FOV
    radius = 20.0 * u.deg

    print("Creating Low-SKA configuration")
    # Chose same coordinate frame for visibilities and healpix map
    phasecentre = SkyCoord(frame="galactic", b=5.0 * u.deg, l=-15.0 * u.deg)
    lowr3 = create_named_configuration("LOWBD2", rmax=3_000.0)

    print("Allocating visibilities")
    vis = create_visibility(
        config=lowr3,
        times=np.linspace(-4, 4, 540) * np.pi / 12,
        frequency=np.r_[15e7],
        channel_bandwidth=np.r_[5e4],
        weight=1.0,
        phasecentre=phasecentre.icrs,
        polarisation_frame=PolarisationFrame("stokesI"),
        times_are_ha=False,
    )

    print("Loading sky GT")
    # Load fits file
    hvox_path = os.sep.join(os.path.abspath(hvox.__file__).split(os.sep)[:-3])
    fits_path = os.path.join(
        hvox_path, "data", "models", "haslam408_dsds_Remazeilles2014.fits"
    )
    hdulist = fits.open(fits_path)
    # Get HPIX metadata
    nside = hdulist[1].header["NSIDE"]
    order = hdulist[1].header["ORDERING"]

    # Create healpix object
    haslam_hpix = HEALPix(nside=nside, order=order, frame=Galactic())
    # get cone of interest in HPIX

    haslam_coi = haslam_hpix.cone_search_lonlat(lon=phasecentre.l, lat=phasecentre.b, radius=radius)
    # get center of the cone (for psf)
    haslam_center = hp.pixelfunc.ang2pix(nside, phasecentre.l.value, phasecentre.b.value, nest=False, lonlat=True)

    # get (lon and lat) coordinates of cone of interest
    directions = haslam_hpix.healpix_to_skycoord(haslam_coi)
    directions_center = haslam_hpix.healpix_to_skycoord(haslam_center)

    # convert to lmn
    lmn = np.asarray(skycoord_to_lmn(directions, phasecentre)).T
    lmn_center = np.asarray(skycoord_to_lmn(directions_center, phasecentre)).squeeze()

    # get map values
    haslam_full = hdulist[1].data['temperature'].ravel()

    if sparse_sky:
        # randomly add ones in an empty sky
        haslam_full = hp.read_map(fits_path)
        haslam_full *= 0
        haslam_full[haslam_coi[np.random.permutation(len(haslam_coi))][:1000]] = 1.

    # get map values at cone of interest
    haslam = haslam_full[haslam_coi]

    print("Simulating visibilities")
    vis_estimate = hvox.dirty2vis(uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3),
                                  xyz=lmn.reshape(-1, 3),
                                  dirty=haslam.reshape(-1),
                                  wgt=vis.visibility_acc.flagged_weight.reshape(-1),
                                  normalisation="xyz",
                                  chunked=True)

    print("Estimating dirty image")
    sky_estimate = hvox.vis2dirty(uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3),
                                  xyz=lmn.reshape(-1, 3),
                                  vis=vis_estimate.reshape(-1),
                                  wgt_vis=vis.visibility_acc.flagged_weight.reshape(-1),
                                  normalisation="xyz",
                                  chunked=True)

    print("Estimating PSF")
    psf_estimate = hvox.compute_psf(
        uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3),
        xyz=lmn.reshape(-1, 3),
        xyz_center=lmn_center,
        wgt=vis.visibility_acc.flagged_weight.reshape(-1),
        normalisation="both",
        chunked=True,
    )

    # fig, axs = plt.subplots(
    #     1,
    #     2,
    #     figsize=(10, 4),
    #     subplot_kw={"projection": "3d"},
    # )
    # ids = np.random.permutation(len(lmn))[:1000]
    # axs[0].scatter(*lmn[ids].T, alpha=0.1)
    # axs[0].set_xlabel("x")
    # axs[0].set_ylabel("y")
    # axs[0].set_zlabel("z")
    # uvw = vis.visibility_acc.uvw_lambda.reshape(-1, 3)
    # ids = np.random.permutation(len(uvw))[:1000]
    # axs[1].scatter(*uvw[ids].T, alpha=0.1)
    # axs[1].set_xlabel("u")
    # axs[1].set_ylabel("v")
    # axs[1].set_zlabel("w")
    # plt.show()

    haslam_plot = hp.ma(haslam_full)
    haslam_plot.mask = np.ones_like(haslam_full, dtype=bool)
    haslam_plot.mask[haslam_coi] = False

    dirty_plot = haslam_plot.copy()
    dirty_plot[haslam_coi] = sky_estimate

    psf_plot = haslam_plot.copy()
    psf_plot[haslam_coi] = psf_estimate

    hp.mollview(
        abs(haslam_plot),
        title="Ground Truth",
    )
    hp.graticule()
    plt.show()

    hp.mollview(
        abs(dirty_plot),
        title="Dirty",
    )
    hp.graticule()
    plt.show()

    hp.mollview(
        abs(psf_plot),
        title="PSF",
    )
    hp.graticule()
    plt.show()

    print("Done!")


src/hvox/_nufft.py
import multiprocessing
import warnings

import numpy as np
import pycsou.operator.linop as pycl
import pycsou.util as pycu


def nufft_builder(xyz, vlambda, real, epsilon, nufft_kwargs):
    # Get NUFFT args
    max_mem = nufft_kwargs.get("max_mem", 512)
    nufft_kwargs = dict(
        enable_warnings=nufft_kwargs.get("enable_warnings", True),
        chunked=nufft_kwargs.get("chunked", True),
        parallel=nufft_kwargs.get("parallel", True),
        nthreads=nufft_kwargs.get("nthreads", None),
    )
    if nufft_kwargs["nthreads"] is None:
        div = 4 if nufft_kwargs["chunked"] else 2
        nufft_kwargs.update({"nthreads": multiprocessing.cpu_count() // div})

    nufft = pycl.NUFFT.type3(
        x=xyz, z=2 * np.pi * vlambda, real=real, isign=-1, eps=epsilon, **nufft_kwargs
    )

    if nufft_kwargs["chunked"]:
        # auto-determine a good x/z chunking strategy
        nufft.allocate(*nufft.auto_chunk(max_mem=max_mem))
        xyz_idx, xyz_chunks = nufft.order("x")  # get a good x-ordering
        uvw_lambda_idx, uvw_lambda_chunks = nufft.order("z")  # get a good z-ordering
        nufft = pycl.NUFFT.type3(
            x=xyz[xyz_idx],
            z=2 * np.pi * vlambda[uvw_lambda_idx],
            real=real,
            isign=-1,
            eps=epsilon,
            **nufft_kwargs
        )

        nufft.allocate(xyz_chunks, uvw_lambda_chunks, direct_eval_threshold=1e4)

        if isinstance(xyz_idx, slice):
            xyz_idx = np.arange(xyz_idx.stop)
        if isinstance(uvw_lambda_idx, slice):
            uvw_lambda_idx = np.arange(uvw_lambda_idx.stop - 1)

        sort_func = {
            "dc": {
                "fw": lambda _: _[xyz_idx],
                "bw": lambda _: _[np.argsort(xyz_idx)],
            },
            "vl": {
                "fw": lambda _: _[uvw_lambda_idx],
                "bw": lambda _: _[np.argsort(uvw_lambda_idx)],
            },
        }

    else:
        identity = lambda _: _
        sort_func = {
            "dc": {
                "fw": identity,
                "bw": identity,
            },
            "vl": {
                "fw": identity,
                "bw": identity,
            },
        }

    return nufft, sort_func


def nufft_vis2dirty(
    xyz,
    uvw_lambda,
    visibilities,
    real,
    epsilon,
    nufft_kwargs
):
    nufft, sort_func = nufft_builder(xyz, uvw_lambda, real, epsilon, nufft_kwargs)
    input_data = sort_func["vl"]["fw"](visibilities)
    input_data = pycu.view_as_real(input_data)
    dirty = nufft.adjoint(input_data)
    dirty = sort_func["dc"]["bw"](dirty)
    return dirty


def nufft_dirty2vis(
    xyz,
    uvw_lambda,
    dirty,
    real,
    epsilon,
    nufft_kwargs
):
    nufft, sort_func = nufft_builder(xyz, uvw_lambda, real, epsilon, nufft_kwargs)

    input_data = sort_func["dc"]["fw"](dirty)
    visibilities = pycu.view_as_complex(nufft(input_data))
    visibilities = sort_func["vl"]["bw"](visibilities)
    return visibilities


example_dcos.py
import os

import matplotlib.pyplot as plt
import numpy as np

try:
    import astropy.units as u
    from astropy.coordinates import SkyCoord
    from rascil.processing_components import create_test_image, show_image
    from ska_sdp_datamodels.configuration.config_create import (
        create_named_configuration,
    )
    from ska_sdp_datamodels.science_data_model.polarisation_model import (
        PolarisationFrame,
    )
    from ska_sdp_datamodels.visibility import create_visibility
    from ska_sdp_func_python.util.coordinate_support import skycoord_to_lmn
except:
    raise ValueError("Missing packages. To run example install the additional dependencies (see repository README)")

import hvox

hvox_path = os.sep.join(os.path.abspath(hvox.__file__).split(os.sep)[:-3])
os.environ["RASCIL_DATA"] = os.path.join(hvox_path, "data")


def get_direction_cosines(image):
    _, _, _, npixel = image.pixels.data.shape
    _, _, ny, nx = image["pixels"].shape
    lmesh, mmesh = np.meshgrid(np.arange(ny), np.arange(nx))
    ra_grid, dec_grid = image.image_acc.wcs.sub([1, 2]).wcs_pix2world(lmesh, mmesh, 0)
    ra_grid = np.deg2rad(ra_grid)
    dec_grid = np.deg2rad(dec_grid)
    directions = SkyCoord(
        ra=ra_grid.ravel() * u.rad,
        dec=dec_grid.ravel() * u.rad,
        frame="icrs",
        equinox="J2000",
    )
    l, m, _ = skycoord_to_lmn(directions, image.image_acc.phasecentre)
    jacobian = np.sqrt(1 - l**2 - m**2)
    direction_cosines = np.stack([l, m, jacobian - 1.0], axis=-1).reshape(-1, 3)
    return direction_cosines, jacobian


if __name__ == "__main__":
    print("Allocating visibilities / Creating Low-SKA configuration")
    lowr3 = create_named_configuration("LOWBD2", rmax=3_000.0)
    vis = create_visibility(
        config=lowr3,
        times=np.linspace(-4, 4, 9) * np.pi / 12,
        frequency=np.r_[15e7],
        channel_bandwidth=np.r_[5e4],
        weight=1.0,
        phasecentre=SkyCoord(
            ra=+15.0 * u.deg, dec=-45.0 * u.deg, frame="icrs", equinox="J2000"
        ),
        polarisation_frame=PolarisationFrame("stokesI"),
        times_are_ha=False,
    )

    print("Loading sky GT")
    m31image = create_test_image(
        phasecentre=vis.phasecentre, frequency=np.r_[15e7], cellsize=5e-4
    )
    direction_cosines, jacobian = get_direction_cosines(m31image)
    wgt_dirty = 1 / jacobian.reshape(-1)

    print("Simulating visibilities")
    vis_estimate = hvox.dirty2vis(uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3),
                                  xyz=direction_cosines.reshape(-1, 3),
                                  dirty=m31image.pixels.data.reshape(-1),
                                  wgt_vis=vis.visibility_acc.flagged_weight.reshape(-1),
                                  wgt_dirty=wgt_dirty,
                                  normalisation="xyz",
                                  chunked=True)

    print("Estimating dirty image")
    dirty_estimate = (
        hvox.vis2dirty(uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3),
                       xyz=direction_cosines.reshape(-1, 3),
                       vis=vis_estimate.reshape(-1),
                       wgt_vis=vis.visibility_acc.flagged_weight.reshape(-1),
                       wgt_dirty=wgt_dirty,
                       normalisation="xyz",
                       chunked=True)
        .reshape(m31image.pixels.shape)
        .squeeze()
    )
    print("Estimating PSF")
    psf_estimate = (
        hvox.compute_psf(
            uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3),
            xyz=direction_cosines.reshape(-1, 3),
            wgt_vis=vis.visibility_acc.flagged_weight.reshape(-1),
            wgt_dirty=wgt_dirty,
            normalisation="both",
            chunked=False,
        )
        .reshape(m31image.pixels.shape)
        .squeeze()
    )

    print("Showing results")
    fig, axs = plt.subplots(
        1,
        3,
        figsize=(15, 4),
        subplot_kw={"projection": m31image.image_acc.wcs.sub([1, 2])},
    )
    im1 = axs[0].imshow(
        m31image.pixels.data.squeeze(), origin="lower", cmap="cubehelix"
    )
    axs[0].set_xlabel(m31image.image_acc.wcs.wcs.ctype[0])
    axs[0].set_ylabel(m31image.image_acc.wcs.wcs.ctype[1])
    axs[0].set_title("Sky (ground truth)")
    plt.colorbar(im1, ax=axs[0])
    im2 = axs[1].imshow(psf_estimate, origin="lower", cmap="cubehelix")
    axs[1].set_xlabel(m31image.image_acc.wcs.wcs.ctype[0])
    axs[1].set_ylabel(m31image.image_acc.wcs.wcs.ctype[1])
    axs[1].set_title("Estimated PSF")
    plt.colorbar(im2, ax=axs[1])
    im3 = axs[2].imshow(dirty_estimate, origin="lower", cmap="cubehelix")
    axs[2].set_xlabel(m31image.image_acc.wcs.wcs.ctype[0])
    axs[2].set_ylabel(m31image.image_acc.wcs.wcs.ctype[1])
    axs[2].set_title("Dirty")
    plt.colorbar(im3, ax=axs[2])
    plt.show()

    print("Done!")

    from ska_sdp_func_python.imaging import invert_ng, predict_ng

    vis_ng = vis.copy(deep=True)
    vis_ng = predict_ng(vis_ng, m31image)
    vis_estimate = hvox.dirty2vis(uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3),
                                  xyz=direction_cosines.reshape(-1, 3), dirty=m31image.pixels.data.reshape(-1), normalisation=None, chunked=True)

    plt.scatter(vis_ng.vis.data.ravel(), vis_estimate.ravel())
    plt.xlabel("NG"), plt.ylabel("HVOX")
    plt.show()

    dirty_ng = m31image.copy(deep=True)
    dirty_ng = invert_ng(vis_ng, dirty_ng, dopsf=False, normalise=True)
    psf_ng = m31image.copy(deep=True)
    psf_ng = invert_ng(vis_ng, psf_ng, dopsf=True, normalise=True)
    dirty_estimate = (
        hvox.vis2dirty(uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3), xyz=direction_cosines.reshape(-1, 3),
                       vis=vis_estimate.reshape(-1), wgt_vis=vis.visibility_acc.flagged_weight.reshape(-1),
                       normalisation="uvw", chunked=True)
        .reshape(m31image.pixels.shape)
        .squeeze()
    )

    psf_estimate = (
        hvox.compute_psf(
            uvw_l=vis.visibility_acc.uvw_lambda.reshape(-1, 3),
            xyz=direction_cosines.reshape(-1, 3),
            wgt_vis=vis.visibility_acc.flagged_weight.reshape(-1),
            wgt_dirty=wgt_dirty,
            normalisation="uvw",
            chunked=True,
        )
        .reshape(m31image.pixels.shape)
        .squeeze()
    )

    fig, axs = plt.subplots(
        2,
        2,
        figsize=(10, 10),
        subplot_kw={"projection": m31image.image_acc.wcs.sub([1, 2])},
    )
    im1 = axs[0, 0].imshow(
        dirty_ng[0].pixels.data.squeeze(), origin="lower", cmap="cubehelix"
    )
    axs[0, 0].set_xlabel(m31image.image_acc.wcs.wcs.ctype[0])
    axs[0, 0].set_ylabel(m31image.image_acc.wcs.wcs.ctype[1])
    axs[0, 0].set_title("DIRTY NG")
    plt.colorbar(im1, ax=axs[0, 0])
    im2 = axs[0, 1].imshow(
        psf_ng[0].pixels.data.squeeze(), origin="lower", cmap="cubehelix"
    )
    axs[0, 1].set_xlabel(m31image.image_acc.wcs.wcs.ctype[0])
    axs[0, 1].set_ylabel(m31image.image_acc.wcs.wcs.ctype[1])
    axs[0, 1].set_title("PSF NG")
    plt.colorbar(im2, ax=axs[0, 1])

    im1 = axs[1, 0].imshow(dirty_estimate, origin="lower", cmap="cubehelix")
    axs[1, 0].set_xlabel(m31image.image_acc.wcs.wcs.ctype[0])
    axs[1, 0].set_ylabel(m31image.image_acc.wcs.wcs.ctype[1])
    axs[1, 0].set_title("DIRTY HVOX")
    plt.colorbar(im1, ax=axs[1, 0])
    im2 = axs[1, 1].imshow(psf_estimate, origin="lower", cmap="cubehelix")
    axs[1, 1].set_xlabel(m31image.image_acc.wcs.wcs.ctype[0])
    axs[1, 1].set_ylabel(m31image.image_acc.wcs.wcs.ctype[1])
    axs[1, 1].set_title("PSF HVOX")
    plt.colorbar(im2, ax=axs[1, 1])
    plt.show()
